### PR TITLE
Add plurals support to Xliff extractor

### PR DESCRIPTION
**Pull request by nicks**
Partial implementation of https://github.com/google/closure-templates/issues/123

### CODE REVIEW

**Comment by mikesamuel** [comment]
I'll patch this in and report back.

File: java/src/com/google/template/soy/xliffmsgplugin/XliffGenerator.java [hunk 1]
```diff
@@ -90,63 +105,129 @@ static CharSequence generateXliff(
...
+   */
+  private void generateSoyMsgHelper(IndentedLinesBuilder ilb, SoyMsg msg, List<SoyMsgPart> parts, String idSuffix) {
+    SoyMsgPart firstPart = Iterables.getFirst(parts, null);
+    if (firstPart instanceof SoyMsgSelectPart) {
+      if (parts.size() != 1) {
+        throw new IllegalStateException("Illegal SELECT AST");
       }
-      ilb.appendLineEnd(">");
+
+      ilb.appendLine("<group restype=\"" + XliffResType.SELECT + "\">");
```
Comment (mikesamuel) [inline java/src/com/google/template/soy/xliffmsgplugin/XliffGenerator.java]: I'm not familiar with the xliff spec.  Why are selects and plurals done this way?

File: java/src/com/google/template/soy/xliffmsgplugin/XliffGenerator.java [hunk 2]
```diff
@@ -40,7 +46,19 @@
...
 class XliffGenerator {
 
-  private XliffGenerator() {}
+  private final Escaper attributeEscaper = XmlEscapers.xmlAttributeEscaper();
+  private final Escaper contentEscaper = XmlEscapers.xmlContentEscaper();
+  private final String sourceLocaleString;
+  private final String targetLocaleString;
+  private final boolean hasTarget;
+
+  private final IndentedLinesBuilder ilb = new IndentedLinesBuilder(2);
```
Comment (mikesamuel) [inline java/src/com/google/template/soy/xliffmsgplugin/XliffGenerator.java]: Does anything bad happen if this allocation happens inside generate?

File: java/src/com/google/template/soy/xliffmsgplugin/XliffGenerator.java [hunk 3]
```diff
@@ -90,63 +105,129 @@ static CharSequence generateXliff(
...
-        }
-        ilb.appendParts(" datatype=\"", attributeEscaper.escape(xliffDatatype), "\"");
+    ilb.decreaseIndent();
+    ilb.appendLine("</body>");
+    ilb.decreaseIndent();
+    ilb.appendLine("</file>");
+    ilb.decreaseIndent();
+    ilb.appendLine("</xliff>");
+
+    return ilb;
```
Comment (mikesamuel) [inline java/src/com/google/template/soy/xliffmsgplugin/XliffGenerator.java]: Does anything bad happen if we return `ilb.toString()`?

File: java/src/com/google/template/soy/xliffmsgplugin/XliffResType.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.template.soy.xliffmsgplugin;
+
+enum XliffResType {
+  SELECT("soy-select"),
+  PLURAL("soy-plural");
```
Comment (mikesamuel) [inline java/src/com/google/template/soy/xliffmsgplugin/XliffResType.java]: http://docs.oasis-open.org/xliff/v1.2/os/xliff-core.html#restype says

> In addition, user-defined values can be used with this attribute. A user-defined value must start with an "x-" prefix.

Should these be `x-soy-...`?

**Comment by nicks** [comment]
Thanks for the review.

Sadly, we cancelled the project that I was building this support for. I'm going to withdraw this PR.

If someone in the future finds this and is interested in adding the soy-plurals/xliff interop, I'm happy to consult on work left to be done. This PR is only about half of it -- the other half is adding the code that reads the XLF file with the translations, and converts it to Closure Templates' internal format.

**Pull request closed without merging**